pkg/host-timesync-daemon/main.go

```diff
@@ -0,0 +1,120 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"syscall"
+	"time"
+	"unsafe"
+
+	"github.com/linuxkit/virtsock/pkg/vsock"
+)
+
+// Listen for connections on an AF_HVSOCK address and update the system time
+// from the hardware clock when a connection is received.
+
+// From <linux/rtc.h> struct rtc_time
+type rtcTime struct {
+	tmSec   uint32
+	tmMin   uint32
+	tmHour  uint32
+	tmMday  uint32
+	tmMon   uint32
+	tmYear  uint32
+	tmWday  uint32
+	tmYday  uint32
+	tmIsdst uint32
+}
+
+const (
+	// iocREAD and friends are from <linux/asm-generic/ioctl.h>
+	iocREAD      = uintptr(2)
+	iocNRBITS    = uintptr(8)
+	iocNRSHIFT   = uintptr(0)
+	iocTYPEBITS  = uintptr(8)
+	iocTYPESHIFT = iocNRSHIFT + iocNRBITS
+	iocSIZEBITS  = uintptr(14)
+	iocSIZESHIFT = iocTYPESHIFT + iocTYPEBITS
+	iocDIRSHIFT  = iocSIZESHIFT + iocSIZEBITS
+	// rtcRDTIMENR and friends are from <linux/rtc.h>
+	rtcRDTIMENR   = uintptr(0x09)
+	rtcRDTIMETYPE = uintptr(112)
+)
+
+func rtcReadTime() rtcTime {
+	f, err := os.Open("/dev/rtc0")
+	if err != nil {
+		log.Fatalf("Failed to open /dev/rtc0: %v", err)
+	}
+	defer f.Close()
+	result := rtcTime{}
+	arg := uintptr(0)
+	arg |= (iocREAD << iocDIRSHIFT)
+	arg |= (rtcRDTIMETYPE << iocTYPESHIFT)
+	arg |= (rtcRDTIMENR << iocNRSHIFT)
+	arg |= (unsafe.Sizeof(result) << iocSIZESHIFT)
+	_, _, errno := syscall.Syscall(syscall.SYS_IOCTL, f.Fd(), arg, uintptr(unsafe.Pointer(&result)))
+	if errno != 0 {
+		log.Fatalf("RTC_RD_TIME failed: %v", errno)
+	}
+	return result
+}
+
+func main() {
+	// host-timesync-daemon -cid <cid> -port <port>
+
+	cid := flag.Int("cid", 0, "AF_HVSOCK CID to listen on")
+	port := flag.Int("port", 0, "AF_HVSOCK port to listen on")
```

Comment:
I think both those lines should say `AF_VSOCK` not `HVSOCK` (especially as hvsock is going away)